File: database-service/src/main/java/com/bnorm/barkeep/db/DbRecipeService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db;

import java.util.Collection;
import java.util.Collections;
import java.util.List;

import javax.persistence.EntityManager;

import com.bnorm.barkeep.model.Ingredient;
import com.bnorm.barkeep.model.Recipe;
import com.bnorm.barkeep.model.User;
import com.bnorm.barkeep.service.RecipeService;


public class DbRecipeService implements RecipeService {

  private final EntityManager em;

  public DbRecipeService(EntityManager entityManager) {
    this.em = entityManager;
  }

  private UserEntity entity(User user) {
    return em.find(UserEntity.class, user.getId());
  }

  private IngredientEntity entity(Ingredient ingredient) {
    return em.find(IngredientEntity.class, ingredient.getId());
  }

  @Override
  public Collection<Recipe> listRecipes() {
    List<RecipeEntity> recipeEntities = em.createNamedQuery("RecipeEntity.findAll", RecipeEntity.class).getResultList();
    return Collections.unmodifiableList(recipeEntities);
  }

  @Override
  public RecipeEntity getRecipe(long recipeId) {
    return em.find(RecipeEntity.class, recipeId);
  }

  @Override
  public RecipeEntity createRecipe(Recipe recipe) {
    if (recipe.getId() != null) {
      throw new IllegalArgumentException(String.format("Cannot create recipe that already has an id=%d",
                                                       recipe.getId()));
    }

    RecipeEntity recipeEntity = new RecipeEntity();
    UserEntity userEntity = entity(recipe.getOwner());

    recipeEntity.setTitle(recipe.getTitle());
    recipeEntity.setDescription(recipe.getDescription());
    recipeEntity.setOwner(userEntity);
    recipeEntity.setImageUrl(recipe.getImageUrl());
    recipeEntity.setInstructions(recipe.getInstructions());
    recipeEntity.setSource(recipe.getInstructions());

    for (com.bnorm.barkeep.model.Component component : recipe.getComponents()) {
      ComponentEntity componentEntity = new ComponentEntity();
      componentEntity.setIngredient(entity(component.getIngredient()));
      componentEntity.setMin(component.getMin());
      componentEntity.setMax(component.getMax());
      componentEntity.setOrder(component.getOrder());
      componentEntity.setComponentNum(component.getComponentNum());
      recipeEntity.addComponent(componentEntity);
    }

    em.getTransaction().begin();
    em.persist(recipeEntity);
    em.getTransaction().commit();
    return recipeEntity;
  }

  @Override
  public RecipeEntity setRecipe(long recipeId, Recipe recipe) {
    RecipeEntity recipeEntity = getRecipe(recipeId);
    if (recipeEntity == null) {
      throw new IllegalArgumentException(String.format("Cannot find recipe with id=%d", recipeId));
    }
    if (recipe.getId() != null && recipe.getId() != recipeId) {
      throw new IllegalArgumentException(String.format("Cannot update recipe with a different id=%d then existing id=%d",
                                                       recipe.getId(),
                                                       recipeId));
    }

    if (recipe.getTitle() != null) {
      recipeEntity.setTitle(recipe.getTitle());
    }
    if (recipe.getDescription() != null) {
      recipeEntity.setDescription(recipe.getTitle());
    }
    if (recipe.getImageUrl() != null) {
      recipeEntity.setImageUrl(recipe.getImageUrl());
    }
    if (recipe.getInstructions() != null) {
      recipeEntity.setInstructions(recipe.getInstructions());
    }
    if (recipe.getSource() != null) {
      recipeEntity.setSource(recipe.getSource());
    }
    if (recipe.getComponents() != null) {
      // todo is this right?
      for (com.bnorm.barkeep.model.Component component : recipe.getComponents()) {
        ComponentEntity componentEntity = new ComponentEntity();
        componentEntity.setIngredient(entity(component.getIngredient()));
        componentEntity.setMin(component.getMin());
        componentEntity.setMax(component.getMax());
        componentEntity.setOrder(component.getOrder());
        componentEntity.setComponentNum(component.getComponentNum());
        recipeEntity.addComponent(componentEntity);
      }
    }

    em.getTransaction().begin();
    em.merge(recipeEntity);
    em.getTransaction().commit();
    return recipeEntity;
  }

  @Override
  public void deleteRecipe(long recipeId) {
    RecipeEntity recipeEntity = getRecipe(recipeId);
    if (recipeEntity == null) {
      throw new IllegalArgumentException(String.format("Cannot find recipe with id=%d", recipeId));
    }
    em.remove(recipeEntity);
  }
}

File: service-database/src/main/kotlin/com/bnorm/barkeep/service/db/DbRecipeService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.service.db

import com.bnorm.barkeep.model.Recipe
import com.bnorm.barkeep.model.RecipeSpec
import com.bnorm.barkeep.model.db.ComponentEntity
import com.bnorm.barkeep.model.db.RecipeEntity
import com.bnorm.barkeep.service.RecipeService
import javax.persistence.EntityManager

class DbRecipeService(private val emPool: Pool<EntityManager>,
                      private val ingredientService: DbIngredientService,
                      private val userService: DbUserService) : RecipeService {

  override fun getRecipes(): List<Recipe> {
    emPool.borrow {
    return createNamedQuery("RecipeEntity.findAll", RecipeEntity::class.java).resultList
    }
    return emptyList() // never used
  }

  fun findRecipe(id: Long): RecipeEntity? {
    emPool.borrow {
      return find(RecipeEntity::class.java, id)
    }
    return null
  }

  override fun getRecipe(id: Long): RecipeEntity {
    return findRecipe(id) ?: throw IllegalArgumentException("Cannot find recipe with id=$id")
  }

  override fun createRecipe(recipe: RecipeSpec): RecipeEntity {
    val owner = recipe.owner ?: throw IllegalArgumentException("Cannot create recipe without an owner")
    val userEntity = userService.findUser(owner.id) ?: throw IllegalArgumentException("Cannot create recipe with an unknown owner id=${owner.id}")

    val recipeEntity = RecipeEntity()

    recipeEntity.title = recipe.title
    recipeEntity.description = recipe.description
    recipeEntity.owner = userEntity
    recipeEntity.imageUrl = recipe.imageUrl
    recipeEntity.instructions = recipe.instructions
    recipeEntity.source = recipe.instructions
    recipeEntity.addUser(userEntity)

    val components = recipe.components
    if (components != null) {
      for (component in components) {
        val componentEntity = ComponentEntity()
        componentEntity.ingredient = ingredientService.getIngredient(component.ingredient.id)
        componentEntity.min = component.min
        componentEntity.max = component.max
        componentEntity.order = component.order
        componentEntity.componentNum = component.componentNum
        recipeEntity.addComponent(componentEntity)
      }
    }

    emPool.txn {
      persist(recipeEntity)
      userEntity.addRecipe(recipeEntity)
    }
    return recipeEntity
  }

  override fun setRecipe(id: Long, recipe: RecipeSpec): RecipeEntity {
    val recipeEntity = getRecipe(id)

    emPool.txn {
      recipe.title?.apply { recipeEntity.title = this }
      recipe.description?.apply { recipeEntity.description = this }
      recipe.owner?.apply { recipeEntity.owner = userService.getUser(this.id) }
      recipe.imageUrl?.apply { recipeEntity.imageUrl = this }
      recipe.instructions?.apply { recipeEntity.instructions = this }
      recipe.source?.apply { recipeEntity.source = this }
      recipe.components?.apply {
        // todo is this right?
        for (component in this) {
          val componentEntity = ComponentEntity()
          componentEntity.ingredient = ingredientService.getIngredient(component.ingredient.id)
          componentEntity.min = component.min
          componentEntity.max = component.max
          componentEntity.order = component.order
          componentEntity.componentNum = component.componentNum
          recipeEntity.addComponent(componentEntity)
        }
      }
    }
    return recipeEntity
  }

  override fun deleteRecipe(id: Long) {
    emPool.txn {
      remove(getRecipe(id))
    }
  }
}
